pkg/controller/direct/datacatalog/tag_controller.go

```diff
@@ -0,0 +1,380 @@
+	if err != nil {
+		return fmt.Errorf("creating datacatalog tag for parent %q: %w", parentName, err)
+	}
+	log.V(2).Info("successfully created datacatalog tag in gcp", "name", created.Name)
+
+	status := &krm.DataCatalogTagStatus{}
+	status.ObservedState = DataCatalogTagObservedState_FromProto(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	// Store the generated name for future Find/Update/Delete operations.
+	status.ExternalRef = direct.LazyPtr(created.Name)
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+// Update implements directbase.Adapter.
+// Only the 'fields' map is mutable. Template and scope (column) are immutable.
+func (a *tagAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	log := klog.FromContext(ctx)
+	mapCtx := &direct.MapContext{}
+
+	// The name (including the system-generated ID) is required for update.
+	if a.actual == nil || a.actual.Name == "" {
+		// This should not happen if Find ran successfully.
+		// Could also fetch from status.ExternalRef if needed.
+		return fmt.Errorf("cannot update tag: actual state or name not found")
+	}
+	tagName := a.actual.Name
+	log.V(2).Info("updating datacatalog tag", "name", tagName)
+
+	desiredProto := DataCatalogTagSpec_ToProto(mapCtx, &a.desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	// Set required fields for the update request payload
+	desiredProto.Name = tagName
+	desiredProto.Template = a.actual.Template // Template is immutable, use actual value
+
+	// Check if only mutable fields ('fields') have changed.
+	updateMask := &fieldmaskpb.FieldMask{}
+	if !reflect.DeepEqual(desiredProto.Fields, a.actual.Fields) {
+		updateMask.Paths = append(updateMask.Paths, "fields")
+	}
+
+	// Check for immutable field changes (should be rejected by KCC webhook ideally)
+	if desiredProto.GetColumn() != a.actual.GetColumn() {
+		return fmt.Errorf("attempting to update immutable field 'column' for tag %q", tagName)
+	}
+	// No need to check 'template' as we force it to the actual value above.
+
+	var updated *pb.Tag
+	var err error
+	if len(updateMask.Paths) == 0 {
+		log.V(2).Info("no mutable field needs update", "name", tagName)
+		updated = a.actual // Use current actual state for status update
+	} else {
+		log.V(2).Info("updating fields", "name", tagName, "fields", updateMask.Paths)
+		req := &pb.UpdateTagRequest{
+			Tag:        desiredProto,
+			UpdateMask: updateMask,
+		}
+		updated, err = a.gcpClient.UpdateTag(ctx, req)
+		if err != nil {
+			return fmt.Errorf("updating datacatalog tag %q: %w", tagName, err)
+		}
+		log.V(2).Info("successfully updated datacatalog tag in gcp", "name", tagName)
+	}
+
+	status := &krm.DataCatalogTagStatus{}
+	status.ObservedState = DataCatalogTagObservedState_FromProto(mapCtx, updated)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	// Ensure ExternalRef is preserved/updated if somehow missing
+	if status.ExternalRef == nil {
```

Comment:
No action needed but because externalRef is required to be the same, the nil check can actually be skipped.